sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java

```diff
@@ -736,17 +736,21 @@ public void deleteDataset(String projectId, String datasetId)
                         try {
                           return insert.execute().getInsertErrors();
                         } catch (IOException e) {
-                          if (new ApiErrorExtractor().rateLimited(e)) {
+                          if (ApiErrorExtractor.INSTANCE.rateLimited(e)) {
                             LOG.info("BigQuery insertAll exceeded rate limit, retrying");
-                            try {
-                              sleeper.sleep(backoff1.nextBackOffMillis());
-                            } catch (InterruptedException interrupted) {
-                              throw new IOException(
-                                  "Interrupted while waiting before retrying insertAll");
-                            }
+                          } else if (ApiErrorExtractor.INSTANCE
+                              .getErrorMessage(e)
+                              .startsWith("Quota exceeded")) {
```

Comment:
Maybe we should just retry on all errors, instead of trying to match the error message?

Reply:
AFAIK, the worker will fail and retry on all other errors after ten seconds anyway. The question here is whether the given error needs to be silently (no explicit error log) retried with exponential backoff or not. I think it makes sense to use exponential backoff for `quota exceeded` and `rate limit exceeded` errors since they are temporal and there's a high chance of getting resolved by themselves in next few retrials. I'm not sure the same holds true for other possible errors like `field size too large`, `unauthorized`, or `user project missing`.

Reply:
Failing is very expensive. If we throw an exception, not only this record, but all other records in the bundle will be retried. Realistically those other errors will be retried anyway (as the whole work item will be retried).

My biggest concern is that parsing text out of the error message seems very brittle. Rather than risk problems there, seems better to just retry all errrrs here.

Reply:
Error code rateLimitExceeded is well documented and probably it will be less brittle to check for that instead of text "Quota exceeded".
https://cloud.google.com/bigquery/troubleshooting-errors

Also, I think the main concern so far has been Beam sending large number of messages to BQ even after BQ service raises quota exceeded errors. I think this will be somewhat exacerbated by introducing exponential backoff here (more messages before the 10 second failed workitem wait) so this has to be combined with a solution where we perform exponential backoff across all BQ streaming write threads started by a given workitme (which can be a separate PR).

Reply:
Sorry I meant error code quotaExceeded.

Reply:
Where do you get a 10-second wait for failed workitem?

Reply:
@rangadi should know the exact place. I think we inject a 10 second wait before re-trying a failed workitem for Dataflow streaming.

Reply:
> My biggest concern is that parsing text out of the error message seems very brittle. Rather than risk problems there, seems better to just retry all errrrs here.

ApiErrorExtractor doesn't have a method for determining if the given exception indicates 'quota exceeded'.

https://github.com/GoogleCloudPlatform/bigdata-interop/blob/master/util/src/main/java/com/google/cloud/hadoop/util/ApiErrorExtractor.java

Is there any other class we can safely extract such information?

Reply:
@chamikaramj That would be runner specific, and I don't believe that's actually true for any runner. We can talk offline.